File: c/src/treenode.c
#include <stddef.h>
#include <stdlib.h>
#include "treenode.h"

struct TreeNode *create_treenode(int val)
{
	struct TreeNode *node = (struct TreeNode *)malloc(sizeof(struct TreeNode));
	node->val = val;
	node->left = NULL;
	node->right = NULL;
	return node;
}

int treenodes_equal(struct TreeNode *t1, struct TreeNode *t2)
{
	if (t1 == NULL && t2 == NULL)
	{
		return 1;
	}
	if (t1 == NULL || t2 == NULL)
	{
		return 0;
	}
	if (t1->val != t2->val)
	{
		return 0;
	}
	return treenodes_equal(t1->left, t2->left) && treenodes_equal(t1->right, t2->right);
}

File: kotlin/src/main/kotlin/TreeNode.kt
package o.mg

class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null


    override fun toString(): String {
        if (this == null) {
            return "null"
        }
        return "<{${this.`val`}}, $left, $right>"
    }

    override fun equals(other: Any?): Boolean {
        println("$this, $other")
        if (other !is TreeNode?) {
            return false
        }
        if (this == null && other == null) return true
        if (this == null || other == null) return false
        if (this === other) return true
        if (this.`val` != other!!.`val`) return false
        return this.left == other.left && this.right == other.right
    }
}

